core/src/main/java/org/apache/iceberg/IncrementalDataTableScan.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import java.util.Collection;
+import java.util.List;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.util.SnapshotUtil;
+
+class IncrementalDataTableScan extends DataTableScan {
+  private Long fromSnapshotId;
+  private Long toSnapshotId;
+
+  IncrementalDataTableScan(TableOperations ops, Table table, Schema schema,
+                           Expression rowFilter, boolean caseSensitive, boolean colStats,
+                           Collection<String> selectedColumns, ImmutableMap<String, String> options,
+                           Long fromSnapshotId, Long toSnapshotId) {
+    super(ops, table, null, schema, rowFilter, caseSensitive, colStats, selectedColumns, options);
+    Preconditions.checkArgument(fromSnapshotId != toSnapshotId, "fromSnapshotId and toSnapshotId cannot be the same");
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+  }
+
+  @Override
+  public TableScan asOfTime(long timestampMillis) {
+    throw new UnsupportedOperationException("asOfTime() is not supported in IncrementalScan");
```

Comment:
The caller doesn't know what scan objects Iceberg uses internally because these are package-private. So it would be strange for Iceberg to throw this error message for a call chain like this:

```java
table.newScan()
    .appendsAfter(lastProcessedId)
    .filter(Expressions.in("id", 1, 2, 3))
    .select("id", "data")
    .asOfTime(System.currentTimeMillis())
    .planFiles();
// => asOfTime() is not supported in IncrementalScan
```

This begs the question "what's an IncrementalScan?" Rather than assuming that the caller knows that it is an incremental scan once `appendsAfter` is called, let's update the error message to be more specific about what went wrong: "Cannot use table state as of time (%s) when processing incremental data in snapshots (%s, %s]"